rpcs3/Crypto/unedat.cpp

```diff
@@ -433,15 +433,25 @@ int check_data(unsigned char *key, EDAT_HEADER *edat, NPD_HEADER *npd, const fs:
 			edat_log.warning("COMPRESSED data detected!");
 	}
 
-	const int block_num = static_cast<int>((edat->file_size + edat->block_size - 1) / edat->block_size);
-	const int metadata_offset = 0x100;
-	const int metadata_size = metadata_section_size * block_num;
+	if (!edat->block_size)
+	{
+		return 1;
+	}
+
+	const usz block_num = utils::aligned_div<u64>(edat->file_size, edat->block_size);
+	const usz metadata_offset = 0x100;
+	const usz metadata_size = utils::mul_saturate<u64>(metadata_section_size, block_num);
 	u64 metadata_section_offset = metadata_offset;
 
-	long bytes_read = 0;
-	long bytes_to_read = metadata_size;
-	std::unique_ptr<u8[]> metadata(new u8[metadata_size]);
-	std::unique_ptr<u8[]> empty_metadata(new u8[metadata_size]);
+	if (utils::add_saturate<u64>(utils::add_saturate<u64>(file_offset, metadata_section_offset), metadata_size) > f->size())
+	{
+		return 1;
+	}
+
+	u64 bytes_read = 0;
+	u64 bytes_to_read = metadata_size;
```

Comment:
Remove this variable and change the check below to:
while (bytes_read < metadata_size)

Reply:
Done.